// freemarker/dist/polyfill/assign-json.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.cleanFile = exports.createTmp = undefined;

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var createTmp = exports.createTmp = function () {
  var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee2(p1, data, tagSyntax) {
    var _this = this;

    return _regenerator2.default.wrap(function _callee2$(_context2) {
      while (1) {
        switch (_context2.prev = _context2.next) {
          case 0:
            return _context2.abrupt('return', new _promise2.default(function () {
              var _ref2 = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(resolve, reject) {
                var _path$parse, name, dir, _tempPath, _tpl, lines, _res;

                return _regenerator2.default.wrap(function _callee$(_context) {
                  while (1) {
                    switch (_context.prev = _context.next) {
                      case 0:
                        _path$parse = path.parse(p1), name = _path$parse.name, dir = _path$parse.dir;
                        _tempPath = path.join(dir, '__temp__' + name + '.ftl');
                        _tpl = reduceMockTpl(data, null, tagSyntax);
                        lines = _tpl.length;
                        _context.prev = 4;
                        _context.t0 = _tpl;
                        _context.next = 8;
                        return fs.readFile(p1);

                      case 8:
                        _context.t1 = _context.sent;

                        _context.t0.push.call(_context.t0, _context.t1);

                        _res = fs.writeFile(_tempPath, _tpl.join('\n'));

                        resolve({
                          tempPath: _tempPath,
                          lines: lines,
                          cleanFile: function cleanFile() {
                            _cleanFile(_tempPath);
                          }
                        });
                        _context.next = 17;
                        break;

                      case 14:
                        _context.prev = 14;
                        _context.t2 = _context['catch'](4);
                        return _context.abrupt('return', resolve({
                          error: _context.t2
                        }));

                      case 17:
                      case 'end':
                        return _context.stop();
                    }
                  }
                }, _callee, _this, [[4, 14]]);
              }));

              return function (_x4, _x5) {
                return _ref2.apply(this, arguments);
              };
            }()));

          case 1:
          case 'end':
            return _context2.stop();
        }
      }
    }, _callee2, this);
  }));

  return function createTmp(_x, _x2, _x3) {
    return _ref.apply(this, arguments);
  };
}();

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var fs = require('../util/fs');
var path = require('path');

var escapeSymbol = function escapeSymbol(str) {
  return str.replace(/((\#\{)|(\@\{)|(\$\{))([^}]*)\}?/g, function () {
    return arguments.length <= 5 ? undefined : arguments[5];
  });
};

function reduceMockTpl(mockData, tpl, tagSyntax) {
  return (0, _keys2.default)(mockData).filter(function (item) {
    return !~item.indexOf('.');
  }).map(function (item) {
    var _value = escapeSymbol((0, _stringify2.default)(mockData[item]));
    var ftlAssign = '<#assign ' + item + ' = ' + _value + '/>';
    if (tagSyntax === 'squareBracket') {
      ftlAssign = ftlAssign.replace('<', '[').replace('>', ']');
    }
    return ftlAssign;
  });
}

;

function _cleanFile(tempPath) {
  fs.delFile(tempPath);
}
exports.cleanFile = _cleanFile;
