// freemarker/dist/app.js
'use strict';

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _entries = require('babel-runtime/core-js/object/entries');

var _entries2 = _interopRequireDefault(_entries);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var os = require('os');
var fs = require('fs');
var path = require('path');
var crypto = require('crypto');
var execFile = require('child_process').execFile;
var assignJson = require('./polyfill/assign-json');

var Freemarker = function () {
  function Freemarker() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
    (0, _classCallCheck3.default)(this, Freemarker);

    this.tmpDir = os.tmpdir();
    this.sourceRoot = options.root || this.tmpDir;
    this.suffix = '.' + (options.suffix || 'ftl');
    this.tagSyntax = options.tagSyntax || 'angleBracket';
    this.cmd = path.join(path.resolve(__dirname, '..'), 'fmpp/bin/fmpp' + (os.platform() === 'win32' ? '.bat' : ''));
  }

  (0, _createClass3.default)(Freemarker, [{
    key: '_randomFile',
    value: function _randomFile() {
      return path.join(this.tmpDir, crypto.randomBytes(20).toString('hex'));
    }
  }, {
    key: '_writeConfig',
    value: function _writeConfig(configFile) {
      var config = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      var str = '';
      for (var key in config) {
        str += key + ': ' + config[key] + '\n';
      }
      fs.writeFileSync(configFile, str, 'utf8');
    }
  }, {
    key: '_writeData',
    value: function _writeData(tddFile) {
      var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};

      fs.writeFileSync(tddFile, (0, _stringify2.default)(data), 'utf8');
    }
  }, {
    key: '_writeFTL',
    value: function _writeFTL(ftlFile) {
      var str = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : '';

      fs.writeFileSync(ftlFile, str, 'utf8');
    }
  }, {
    key: '_cleanFiles',
    value: function _cleanFiles() {
      var files = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : [];

      files.forEach(function (file) {
        fs.existsSync(file) && fs.unlinkSync(file);
      });
    }
  }, {
    key: '_getRealPath',
    value: function _getRealPath(file) {
      var _file = file;
      if (!_file.endsWith(this.suffix)) {
        _file += this.suffix;
      }
      if (!path.isAbsolute(_file)) {
        _file = path.join(this.sourceRoot, _file);
      }
      return _file;
    }
  }, {
    key: 'render',
    value: function render(str, data, callback) {
      var _this = this;

      var ftlFile = this._randomFile() + this.suffix;
      this._writeFTL(ftlFile, str);
      this.renderFile(ftlFile, data, function (err, result) {
        callback(err, result);
        _this._cleanFiles([ftlFile]);
      });
    }
  }, {
    key: 'renderFile',
    value: function () {
      var _ref = (0, _asyncToGenerator3.default)( /*#__PURE__*/_regenerator2.default.mark(function _callee(file) {
        var data = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
        var callback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : function () {};

        var _file, _ref2, tempPath, cleanFile, error, lines;

        return _regenerator2.default.wrap(function _callee$(_context) {
          while (1) {
            switch (_context.prev = _context.next) {
              case 0:
                _file = this._getRealPath(file);

                if (!((0, _entries2.default)(data).length === 0)) {
                  _context.next = 3;
                  break;
                }

                return _context.abrupt('return', this.renderProxy(_file, {}, callback));

              case 3:
                _context.next = 5;
                return assignJson.createTmp(_file, data, this.tagSyntax);

              case 5:
                _ref2 = _context.sent;
                tempPath = _ref2.tempPath;
                cleanFile = _ref2.cleanFile;
                error = _ref2.error;
                lines = _ref2.lines;

                if (!error) {
                  _context.next = 12;
                  break;
                }

                return _context.abrupt('return', callback(error));

              case 12:
                this.renderProxy(tempPath, {}, function (error, result) {
                  callback(error ? error.replace(/line (\d+)\,/g, function (match, line) {
                    return 'line ' + (Number(line) - lines) + ',';
                  }) : error, result);
                  cleanFile();
                });

              case 13:
              case 'end':
                return _context.stop();
            }
          }
        }, _callee, this);
      }));

      function renderFile(_x6) {
        return _ref.apply(this, arguments);
      }

      return renderFile;
    }()
  }, {
    key: 'renderProxy',
    value: function renderProxy(file, data, callback) {
      var _this2 = this;

      if (!file) return callback('No ftl file');

      var htmlFile = this._randomFile();
      var tddFile = this._randomFile();
      var configFile = this._randomFile();
      var config = {
        sourceRoot: this.sourceRoot,
        tagSyntax: this.tagSyntax,
        outputFile: htmlFile,
        sourceEncoding: 'UTF-8',
        outputEncoding: 'UTF-8',
        data: 'tdd(' + tddFile + ')'
      };
      this._writeData(tddFile, data);
      this._writeConfig(configFile, config);

      execFile(this.cmd, [file, '-C', configFile], function (err, log) {
        var result = '';
        if (fs.existsSync(htmlFile)) {
          result = fs.readFileSync(htmlFile, 'utf8');
        }
        callback(err || !/DONE/.test(log) ? log : null, result);
        _this2._cleanFiles([htmlFile, tddFile, configFile]);
      });
    }
  }]);
  return Freemarker;
}();

;

module.exports = Freemarker;
